tiledb/cc/array.cc

```diff
@@ -84,20 +110,113 @@ void init_array(py::module &m) {
 
              Array::consolidate(ctx, self.uri(), config);
            })
-      .def("vacuum", &Array::vacuum)
-      .def("create",
-           py::overload_cast<const std::string &, const ArraySchema &>(
-               &Array::create))
-      .def("load_schema",
+      .def("_vacuum", &Array::vacuum)
+      .def("_create",
+           [](const Context &ctx, const std::string &uri,
+              const ArraySchema &schema) {
+             ctx.handle_error(tiledb_array_create(ctx.ptr().get(), uri.c_str(),
+                                                  schema.ptr().get()));
+           })
+      .def("_load_schema",
            py::overload_cast<const Context &, const std::string &>(
                &Array::load_schema))
-      .def("encryption_type", &Array::encryption_type)
+      .def("_encryption_type", &Array::encryption_type)
+
+      .def("_non_empty_domain_is_empty",
+           [](Array &self, const unsigned &dim_idx, const py::dtype &n_type,
+              const Context &ctx) -> bool {
+             int32_t is_empty = 0;
 
-      // TODO non_empty_domain
-      // TODO non_empty_domain_var
+             if (py::getattr(n_type, "kind").is(py::str("S")) ||
+                 py::getattr(n_type, "kind").is(py::str("U"))) {
+               uint64_t start_size, end_size;
+               ctx.handle_error(
+                   tiledb_array_get_non_empty_domain_var_size_from_index(
+                       ctx.ptr().get(), self.ptr().get(), dim_idx, &start_size,
+                       &end_size, &is_empty));
+               return is_empty == 1;
+             } else {
+               void *domain = nullptr;
+               if (n_type.is(py::dtype::of<uint64_t>())) {
+                 domain = new uint64_t[2];
+                 // int64_t also used for datetime64
+               } else if (n_type.is(py::dtype::of<int64_t>()) ||
+                          py::getattr(n_type, "kind").is(py::str("M"))) {
+                 domain = new int64_t[2];
+               } else if (n_type.is(py::dtype::of<uint32_t>())) {
+                 domain = new uint32_t[2];
+               } else if (n_type.is(py::dtype::of<int32_t>())) {
+                 domain = new int32_t[2];
+               } else if (n_type.is(py::dtype::of<uint16_t>())) {
+                 domain = new uint16_t[2];
+               } else if (n_type.is(py::dtype::of<int16_t>())) {
+                 domain = new int16_t[2];
+               } else if (n_type.is(py::dtype::of<uint8_t>())) {
+                 domain = new uint8_t[2];
+               } else if (n_type.is(py::dtype::of<int8_t>())) {
+                 domain = new int8_t[2];
+               } else if (n_type.is(py::dtype::of<double>())) {
+                 domain = new double[2];
+               } else if (n_type.is(py::dtype::of<float>())) {
+                 domain = new float[2];
```

Comment:
You're going to need to call a `delete[]` so that `domain` doesn't leak here.

Reply:
I would suggest implementing a templated function along the lines of

```
template <typename T>
bool _non_empty_domain_is_empty_aux<T>(...) {
    int32_t is_empty = 0;
    T domain [2];
    ctx.handle_error(tiledb_array_get_non_empty_domain_from_index(
                   ctx.ptr().get(), self.ptr().get(), dim_idx, &domain,
                   &is_empty));
    return is_empty == 1;
}
```

Reply:
Special thanks for this. It was my mistake for not noticing the memory leak here.